### PR TITLE
feature: artifact support

### LINKED ISSUE #18: Add a way to attach artifacts to the completed process
If you run an export with ExportDefinitions ;) you'll end up with a file to download.

The link might point to a single file or a group of files, so called "job artifacts". Nomenclature is quite common with these job-based configurations, see for example https://docs.gitlab.com/ee/user/project/pipelines/job_artifacts.html

If the process creates artifacts, the link to download the archive should be made available from the UI.

If the process is deleted, the download probably should too (?)

### PR DESCRIPTION
**Pull request by dkarlovi**
Closes #18

### CODE REVIEW

**Comment by dkarlovi** [comment]
~~@dpfaffenbauer I think this can be merged as is.~~

File: src/ProcessManagerBundle/Migrations/Version20181023141014.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class Version20181023141014 extends AbstractPimcoreMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $processesTable = $schema->getTable("process_manager_processes");
+        $processesTable->addColumn("artifact", "asset", ['notnull' => true]);
```
Comment (dkarlovi) [inline src/ProcessManagerBundle/Migrations/Version20181023141014.php]: Still need to figure out how this works.

File: src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js [hunk 2]
```diff
@@ -207,6 +208,21 @@ pimcore.plugin.processmanager.processes = Class.create({
...
+                {
+                    text : t('processmanager_artifact_download'),
+                    xtype:'actioncolumn',
+                    width:50,
+                    items: [
+                        {
+                            iconCls : 'pimcore_icon_download',
+                            tooltip: t('processmanager_artifact_download'),
+                            handler: function(grid, rowIndex) {
+                                var id = grid.getStore().getAt(rowIndex).get('id');
```
Comment (dkarlovi) [inline src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js]: This doesn't work (yet)

Comment (dkarlovi) [reply src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js]: Also, not all processes have artifacts attached, need to handle it here.

File: src/ProcessManagerBundle/EventListener/ArtifactListener.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        if (false === array_key_exists('artifact', $record['context'])) {
+            return;
+        }
+        
+        /** @var ProcessInterface $process */
+        $process = $record['extra']['process'];
+
+        // ProcessInterface instance must have a relation to its ExecutableInterface instance
+        // TODO: need asset path in Executable config, hardcoded for now
+        $artifactAssetPath = '/exports';
```
Comment (dkarlovi) [inline src/ProcessManagerBundle/EventListener/ArtifactListener.php]: > ProcessInterface instance must have a relation to its ExecutableInterface instance

File: src/ProcessManagerBundle/EventListener/ArtifactListener.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+        $artifactPath = $record['context']['artifact'];
+        $fileHandle = fopen($artifactPath, 'rb', false, FileHelper::getContext());
+
+        $artifact = new Asset();
+        $artifact->setParent(Asset\Service::createFolderByPath($artifactAssetPath));
+        $artifact->setStream($fileHandle);
+        // TODO: name the file something like "<executable name> artifact <current datetime>" or provide an artifact name template in the Executable config
+        $artifact->setFilename(pathinfo($artifactPath, PATHINFO_FILENAME));
```
Comment (dkarlovi) [inline src/ProcessManagerBundle/EventListener/ArtifactListener.php]: > name the file something like "<executable name> artifact <current datetime>" or provide an artifact name template in the Executable config

File: src/ProcessManagerBundle/EventListener/ArtifactListener.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        // TODO: add an event listener which sets Artifact=null (or deletes the Process) if the user deletes this asset
+        $artifact->addMetadata('process_manager.process', 'number', $process->getId());
```
Comment (dkarlovi) [inline src/ProcessManagerBundle/EventListener/ArtifactListener.php]: > add an event listener which sets `artifact=null` (or deletes this `Process`?) if the user deletes this `Asset`

File: src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js [hunk 6]
```diff
@@ -207,6 +208,21 @@ pimcore.plugin.processmanager.processes = Class.create({
                         }
                     ]
                 },
+                {
+                    text : t('processmanager_artifact_download'),
+                    xtype:'actioncolumn',
+                    width:50,
+                    items: [
+                        {
```
Comment (dkarlovi) [inline src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js]: How to skip drawing an icon here if there's no artifact attached?

Comment (dkarlovi) [reply src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js]: @dpfaffenbauer if you can help here, I think this is ready to merge.

Comment (dpfaffenbauer) [reply src/ProcessManagerBundle/Resources/public/pimcore/js/processes.js]: Easy solution: iconCls can be a function as well, and you could just return an empty string. You also need to check in the handler function if there is anything in there. 

Good and harder solution: https://github.com/w-vision/ImportDefinitions/blob/master/src/ImportDefinitionsBundle/Resources/public/pimcore/js/definition/item.js#L610

File: src/ProcessManagerBundle/EventListener/ArtifactListener.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+namespace ProcessManagerBundle\EventListener;
+
+use Pimcore\Event\Model\AssetEvent;
+use Pimcore\File as FileHelper;
+use Pimcore\Model\Asset;
+use ProcessManagerBundle\Model\Process;
+use ProcessManagerBundle\Model\ProcessInterface;
+use ProcessManagerBundle\Monolog\ProcessLogEvent;
+
+/**
```
Comment (dpfaffenbauer) [inline src/ProcessManagerBundle/EventListener/ArtifactListener.php]: please no class comments.

File: src/ProcessManagerBundle/EventListener/ArtifactListener.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        $process = $record['extra']['process'];
+
+        // ProcessInterface instance must have a relation to its ExecutableInterface instance
+        // TODO: need asset path in Executable config, hardcoded for now
+        $artifactAssetPath = '/exports';
+
+        $artifactPath = $record['context']['artifact'];
+        $fileHandle = fopen($artifactPath, 'rb', false, FileHelper::getContext());
+
+        $artifact = new Asset();
```
Comment (dpfaffenbauer) [inline src/ProcessManagerBundle/EventListener/ArtifactListener.php]: Still don't agree with that, the asset should be created from the process that creates the Process (eg. Import Definitions). Not every process has a executable linked to it. Even Import Definitions logs to PM without an executable.

File: src/ProcessManagerBundle/Monolog/ProcessLogEvent.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * @copyright  Copyright (c) 2015-2017 Dominik Pfaffenbauer (https://www.pfaffenbauer.at)
+ * @license    https://github.com/dpfaffenbauer/ProcessManager/blob/master/gpl-3.0.txt GNU General Public License version 3 (GPLv3)
+ */
+
+namespace ProcessManagerBundle\Monolog;
+
+use Symfony\Component\EventDispatcher\Event;
+
+/**
```
Comment (dpfaffenbauer) [inline src/ProcessManagerBundle/Monolog/ProcessLogEvent.php]: pls no class comments

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]

**Comment by dpfaffenbauer** [comment]
also, this won't be merged until the Export Branch in Import Definitions is merged. One step after another. If we merge to fast everything, it gets too confusing.

**Comment by dpfaffenbauer** [comment]
@dkarlovi I changed this PR according to how I feel it should be implemented

**Comment by dkarlovi** [comment]
@dpfaffenbauer you've also removed the part which actually creates the asset.

I know you've said it should be moved elsewhere, but I don't see it has been?

**Comment by dpfaffenbauer** [comment]
@dkarlovi which part do you exactly mean?

**Comment by dkarlovi** [comment]
In the listener. I see you've moved it to ImportDefinitions though, forget it. :+1:

**Comment by dpfaffenbauer** [comment]
Commands are done

**Comment by dpfaffenbauer** [comment]
sorry, comment belonged to the ImportDefinitions PR :D

**Comment by dkarlovi** [comment]
@dpfaffenbauer keeping this all in my head correctly, it's surprising I'm still commenting on Github, not on Gitlab. :)

**Comment by dpfaffenbauer** [comment]
or Gitter or Slack :D

**Review by dpfaffenbauer — APPROVED** [review]

**Pull request merged**